addons/binding/org.openhab.binding.rotelra1x/src/main/java/org/openhab/binding/rotelra1x/handler/RotelRa1xHandler.java

```diff
@@ -0,0 +1,345 @@
+            updateStatus(ThingStatus.ONLINE);
+            serialPort.getOutputStream().write("get_current_power!".getBytes(StandardCharsets.US_ASCII));
+            updateState(CHANNEL_MUTE, OnOffType.OFF);
+            updateState(CHANNEL_BRIGHTNESS, new PercentType(100));
+        }
+    }
+
+    private void disconnect() {
+        if (serialPort != null) {
+            serialPort.close();
+        }
+        serialPort = null;
+    }
+
+    private String readCommand() throws IOException {
+        return readUntil('!', '=');
+    }
+
+    private String readUntil(char terminator) throws IOException {
+        return readUntil(terminator, terminator);
+    }
+
+    private String readUntil(char terminator1, char terminator2) throws IOException {
+        StringBuilder result = new StringBuilder();
+        while (true) {
+            int b = serialPort.getInputStream().read();
+            if (b == -1) {
+                throw new IOException("Connection unexpectedly closed");
+            } else if (b == terminator1 || b == terminator2) {
+                break;
+            } else {
+                result.append((char) b);
+            }
+        }
+        return result.toString();
+    }
+
+    private PercentType readVolume() throws IOException {
+        String volumeString = readUntil('!');
+        int volume;
+        if ("min".equals(volumeString)) {
+            volume = 0;
+        } else if ("max".equals(volumeString)) {
+            volume = maximumVolume;
+        } else {
+            volume = Integer.parseInt(volumeString, 10);
+        }
+        double volumePct = volume * 100.0 / maximumVolume;
+        return new PercentType(BigDecimal.valueOf(Math.round(volumePct)));
+    }
+
+    private PercentType readDimmer() throws IOException {
+        String dimmerString = readUntil('!');
+        int dimmer = Integer.parseInt(dimmerString, 10);
+        // Let's make 100 % the brightest, makes more sense
+        double dimmerPct = 100.0 - (dimmer * 100.0 / 6.0);
+        return new PercentType(BigDecimal.valueOf(Math.round(dimmerPct)));
+    }
+
+    private DecimalType readFrequency() throws IOException {
+        String freqString = readUntil('!');
+        double freq;
+        if ("off".equals(freqString)) {
+            freq = 0.0;
+        } else {
+            freq = Double.parseDouble(freqString);
+        }
+        return new DecimalType(freq);
+    }
+
+    private void powerOnRefresh() {
+        scheduler.submit(() -> {
+            try {
+                send("get_volume!");
+                send("get_current_source!");
+            } catch (IOException | ConfigurationError e) {
+                logger.warn("Failed to request volume and source after powering on.", e);
+            }
+        });
+    }
+
+    private void inputLoop() {
+        while (!exit) {
+            try {
+                if (serialPort == null) {
```

Comment:
This check is redundant as the first check in `connect` already checks if serialPort is null.